src/hydra-queue-runner/hydra-queue-runner.cc

```diff
@@ -449,74 +463,131 @@ bool State::checkCachedFailure(Step::ptr step, Connection & conn)
-                auto conn(dbPool.get());
-                pqxx::work txn(*conn);
-                txn.parameterized
-                    ("update Builds set notificationPendingSince = null where id = $1")
-                    (item.id)
-                    .exec();
-                txn.commit();
+        FILE* toNotify = fdopen(dup(to.writeSide.get()), "w");
+        FILE* fromNotify = fdopen(dup(from.readSide.get()), "r");
+
+        new_pid = startProcess([&]() {
+
+            if (dup2(to.readSide.get(), STDIN_FILENO) == -1) {
+              throw SysError("cannot dup input pipe to stdin");
+            }
+            if (dup2(from.writeSide.get(), STDOUT_FILENO) == -1) {
+              throw SysError("cannot dup output pipe to stdout");
             }
 
-            nrNotificationTimeMs += std::chrono::duration_cast<std::chrono::milliseconds>(now2 - now1).count();
-            nrNotificationsDone++;
+            execlp("hydra-notify", "hydra-notify", NULL);
 
-        } catch (std::exception & e) {
-            nrNotificationsFailed++;
-            printMsg(lvlError, format("notification sender: %1%") % e.what());
-            sleep(5);
-        }
+            throw SysError("cannot start hydra-notify");
+        });
+
+        close(to.readSide.release());
+        close(from.writeSide.release());
+
+        hydra_notify = std::optional<hydra_notify_state> {{
+          .pid = new_pid,
+          .to_notify = std::move(to),
+          .from_notify = std::move(from),
+          .to_notify_stream = toNotify,
+          .from_notify_stream = fromNotify,
+          .buf_notify = NULL
+        }};
+      }
+
+      NotificationItem item;
+      {
+          auto notificationSenderQueue_(notificationSenderQueue.lock());
+          while (notificationSenderQueue_->empty())
+              notificationSenderQueue_.wait(notificationSenderWakeup);
+          item = notificationSenderQueue_->front();
+          notificationSenderQueue_->pop();
+      }
+
+
+      MaintainCount<counter> mc(nrNotificationsInProgress);
+
+      printMsg(lvlChatty, format("sending notification about build %1%") % item.id);
+
+      auto now1 = std::chrono::steady_clock::now();
+
+      std::string payload;
+
+      switch (item.type) {
+        case NotificationItem::Type::BuildStarted:
+          payload = "build-started"s + " "s + std::to_string(item.id);
+          break;
+        case NotificationItem::Type::BuildFinished:
+          payload = "build-finished"s + " "s + std::to_string(item.id);
+          break;
+        case NotificationItem::Type::StepFinished:
+          payload = "step-finished"s + " "s + std::to_string(item.id) + " "s + std::to_string(item.stepNr) + " "s + item.logPath;
+          break;
+      }
+
+      fprintf(hydra_notify->to_notify_stream, "%s\n", payload.c_str());
+      fflush(hydra_notify->to_notify_stream);
+
+      size_t n = 0;
+      int res = getline(&(hydra_notify->buf_notify), &n, hydra_notify->from_notify_stream);
```

Comment:
According to the documentation, the value of `n` is also supposed to be preserved across calls. You should add it to the `hydra_notify` state.